### DISCUSSION TITLE
Search path images in CustomWidgets and pyInstaller

### DISCUSSION THREAD

**Discussion by esbvaGames** [alejandroautalan/pygubu-designer#198] [General]
Hello:
I found one more problem with the Custom-Widgets, it turns out that when doing it without images, there is no error, and when buttons with images are included, and although these images are in my "./Custom" directory, it does not find them, a solution was , copy the icons, in the "c:\python\script" folder, together with pygubu-designer, so that the widget with images can be opened in the designer, they must also be copied together with the script of the main application that uses the widget , pyinstaller still fails, isn't there a way in pygubu to tell it to include a specific directory, to look for the images??

![AMyQecoLF6](https://user-images.githubusercontent.com/5530499/224424052-c7e80d09-b167-48e0-985f-35705d489e9d.png)
![FK3JZ47Wir](https://user-images.githubusercontent.com/5530499/224424089-931ec28c-8636-4f02-9073-eabf00e5e7d9.png)
![A9IKEKJNxJ](https://user-images.githubusercontent.com/5530499/224424128-0daaa849-26b6-466f-bf94-28472071ddc9.png)
![Uq9pPTw1YE](https://user-images.githubusercontent.com/5530499/224424190-3888b3d3-4373-4909-82b0-2660cbc2591e.png)
![eEwH6PeL6K](https://user-images.githubusercontent.com/5530499/224424211-8d9bcce4-0265-4068-8212-f2d580399a7a.png)

**Comment by alejandroautalan** [comment]
> Hello: I found one more problem with the Custom-Widgets, it turns out that when doing it without images, there is no error, and when buttons with images are included, and although these images are in my "./Custom" directory, it does not find them, a solution was , copy the icons, in the "c:\python\script" folder, together with pygubu-designer, so that the widget with images can be opened in the designer, they must also be copied together with the script of the main application that uses the widget , pyinstaller still fails, isn't there a way in pygubu to tell it to include a specific directory, to look for the images??

Hello,
The builder will search for images in the folder configured with ***builder.add_resource_path*** method. If the images are distributed in different folders, you should tell the builder where they are. Do it before creating the main window.
Example:

```python
PROJECT_PATH = pathlib.Path(__file__).parent
PROJECT_UI = PROJECT_PATH / "helloworld.ui"
IMAGES_DIR1 = PROJECT_PATH / "custom" / "images1"
IMAGES_DIR2 = PROJECT_PATH / "custom" / "images2"

...

class DemoApp:
    def __init__(self, master=None):
        # 1: Create a builder and setup resources path (if you have images)
        self.builder = builder = pygubu.Builder()
        builder.add_resource_path(PROJECT_PATH)
        builder.add_resource_path(IMAGES_DIR1)
        builder.add_resource_path(IMAGES_DIR2)

        # 2: Load an ui file
        builder.add_from_file(PROJECT_UI)
```

Another example is [here](https://github.com/alejandroautalan/pygubu-designer/blob/master/examples/integration_with_pyinstaller/myapp.py)

> pyinstaller still fails, isn't there a way in pygubu to tell it to include a specific directory, to look for the images??

Also, you must tell pyinstaller to include the data files (images).  With a spec file, is done like [this](https://github.com/alejandroautalan/pygubu-designer/blob/master/examples/integration_with_pyinstaller/myapp.spec).

```
data_files = [
    ('myapp.ui', '.'),
    ('imgs/*.gif', 'imgs'),
]
```

I do not remember how to set this in the command line.

Try it and let me know.
Regards.

**Comment by esbvaGames** [comment]
If I saw it, but I didn't find an adequate way to work them separately, it is best to work, the UIs, Widgets and builders work them in the same directory, where the application is built, plus a copy of pygubu-designer-py, also in that directory since when it starts, this will be your working directory, if it is changed, the references of the widgets integrated in the ui are lost, main !!!
